res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,879 @@
+    Prime4.rgbCode[deckColors[3]],
+];
+const colDeckDark = [
+    Prime4.rgbCode[deckColors[0] + "Dark"],
+    Prime4.rgbCode[deckColors[1] + "Dark"],
+    Prime4.rgbCode[deckColors[2] + "Dark"],
+    Prime4.rgbCode[deckColors[3] + "Dark"],
+];
+
+// Register '0x9n' as a button press and 'ox8n' as a button release
+components.Button.prototype.isPress = function(channel, control, value, status) {
+    return (status & 0xF0) === 0x90;
+};
+
+Prime4.DeckAssignButton = function(options) {
+    components.Button.call(this, options);
+
+    if (!Number.isInteger(this.deckIndex)) {
+        throw `invalid deckIndex: ${this.deckIndex}`;
+    }
+    if (!(this.toDeck instanceof Prime4.Deck)) {
+        throw "invalid toDeck";
+    }
+
+    const deckSide = (this.deckIndex % 2) === 0 ? "leftDeck" : "rightDeck";
+    if (!(Prime4[deckSide] instanceof Prime4.Deck)) {
+        throw "invalid deckIndex or structure; We expect Prime4.leftDeck and Prime4.rightDeck to be valid decks representing the physical left and right decks";
+    }
+    const isActive = () => {
+        console.log(this.deckIndex, deckSide);
+        console.log(Prime4[deckSide], Prime4[deckSide].currentDeck);
+        console.log(this.toDeck, this.toDeck.currentDeck);
+        return Prime4[deckSide] === this.toDeck;
+    };
+
+    this.trigger = function() {
+        this.output(isActive());
+    };
+
+    this.outValueScale = function(value) {
+        console.log(value);
+        console.log(deckSide);
+        const scaledVal = components.Button.prototype.outValueScale.call(this, value);
+        console.log(scaledVal);
+        return scaledVal;
+    };
+
+    this.input = function(channel, control, value, status, _group) {
+        console.log(Object.keys(this));
+        if (!this.isPress(channel, control, value, status) || isActive()) {
+            return;
+        }
+        Prime4[deckSide].forEachComponent(c => { c.disconnect(); });
+        Prime4[deckSide] = this.toDeck;
+        this.assignmentButtons.forEachComponent(btn => btn.trigger());
+        Prime4[deckSide].forEachComponent(c => { c.connect(); c.trigger(); });
+    };
+
+};
+
+Prime4.DeckAssignButton.prototype = Object.create(components.Button.prototype);
+
+Prime4.EffectUnitEncoderInput = function(_channel, _control, value, _status, _group) {
+    if (value >= 1 && value < 20) {
+        this.inSetParameter(this.inGetParameter() + (value / 100));
+    } else if (value <= 127 && value > 100) {
+        this.inSetParameter(this.inGetParameter() + ((value - 128) / 100));
+    }
+};
+
+Prime4.init = function(_id, _debug) {
+    // Turn off all LEDs
+    midi.sendShortMsg(0x90, 0x75, 0x00);
+
+    const decks = [
+        new Prime4.Deck(1, 4),
+        new Prime4.Deck(2, 5),
+        new Prime4.Deck(3, 4),
+        new Prime4.Deck(4, 5),
+    ];
+
+    // Disconnect all decks at first so they don't fight with each other
+    decks.forEach(deck => deck.forEachComponent(comp => comp.disconnect()));
+
```

Comment:
For some reason, disconnecting the decks doesn't work. Can you include this logging output and post the controller debug log from when the controller mapping initializes?
```suggestion
    // Disconnect all decks at first so they don't fight with each other
    decks.forEach(deck => deck.forEachComponent(comp => {console.log(`disconnecting "${comp.group}, ${comp.inKey}"`); comp.disconnect();}));

```

Reply:
Note: Before adding this code, I noticed that the jog wheel LEDs show the colours of decks 3 and 4, even though they control decks 1 and 2. This is fixed once the decks are switched to 3 and 4, then back to 1 and 2 again.

Reply:
> Note: Before adding this code, I noticed that the jog wheel LEDs show the colours of decks 3 and 4, even though they control decks 1 and 2. This is fixed once the decks are switched to 3 and 4, then back to 1 and 2 again.


Right, that bug makes sense because the 3/4 decks get connected and triggered after the 1/2 decks. Switching decks retriggers the LEDs to be correct. 

> I've included the info in mixxx/res/controllers/DebugNotesForSwiftb0y.txt. FYI The jog wheel LED behaviour still shows decks 3 and 4's layers' colours on startup.

Thank you. It has already proven useful.

Reply:
So now that technically the output of all decks should be disconnected I need to know whether they still respond to changes within mixxx. Can you try to change the LEDs without touching the controller. For example trying to change Sync, Play, etc via the Mixxx GUI should not result in any change of LED output on the controller. Does it make a difference on what deck you are changing the state? Eg does 1/2 respond to output but 3/4 not? Ideally it should not respond to any change within mixxx.

Reply:
Ah yes, so currently the side decks do not respond to changes in the Mixxx GUI on startup, no matter the deck, except for the performance pads. (The middle `mixerStrip`s still respond to changes FYI. Probably not relevant but still noteworthy.)

Reply:
Mhmm, the fact that the performance pads are fighting is weird. We should look into that.